### PR TITLE
Quiz service

### LINKED ISSUE #3: Multiple choice quiz implementation backend
## Parent
#1 

## Description
Implement the multiple choice quiz in the backend according to the concept.

## Acceptance Criteria

- [ ] Multiple choice quizzes crud operations work

### PR DESCRIPTION
**Pull request by PaulBredl**
I implemented the quiz service. Due to previously having this pull request only for the schema, the pull request template is lost but imo we should change the pr template anyways because it's fucking tedious and no one really fills it out completely anyway.

Changes:
- some more changes in the schema to make it easier for the frontend
- implementation
- tests

## How has this been tested?
Please describe the test strategy you followed.

- [ ] automated unit test
- [x] automated integration test
- [ ] automated acceptance test
- [x] manual, exploratory test
    
## Checklist before requesting a review
- [x] My code follows the [coding guidelines](https://github.com/IT-REX-Platform/wiki/blob/main/dev-manuals/backend/coding-guidelines.md) of this project
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] My code fulfills all acceptance criteria
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] The test coverage (line and branch) is reasonably high, especially on Service classes and other classes with complex logic
- [ ] I have made corresponding changes to the documentation
- [ ] I have added explanation of architectural decision and rationales to [wiki/adr](https://github.com/IT-REX-Platform/wiki/tree/main/adr)
- [ ] I have updated the changes in the ticket description

## Checklist for reviewer
- [ ] The code works and does not throw errors
- [ ] The code is easy to understand and there are no confusing parts
- [ ] The code follows the [coding guidelines](https://github.com/IT-REX-Platform/wiki/blob/main/dev-manuals/backend/coding-guidelines.md) of this project
- [ ] The code change accomplishes what it is supposed to do
- [ ] I cannot think of any use case in which the code does not behave as intended
- [ ] The added and existing tests reasonably cover the code change
- [ ] I cannot think of any test cases, input or edge cases that should be tested in addition
- [ ] Description of the change is included in the documentation
## Schema updates
I updated the quiz schema from the original concept in the following ways:
1. Added feedback for each answer. This is an optional text that can be shown to the user if he selected that particular answer.
2. Added the number of the question to the question type. This is used to determine the order of the questions.
3. Added a "poolingMode" which determines if the questions should be shown random or in order. (The actual logic for that will handle the backend)
4. Added a more usable way to update quizzes. This is named "modifyQuiz" instead of "updateQuiz" to differentiate it from our usual update mutations. This query can update quizzes in various ways via subqueries. 

## Example mutation for the modification mutation
```graphql
mutation modifyQuiz {
    modifyQuiz(id: "5f9b3b3b9b0b1e0b3c9b0b1e") {
        # update some metadata
        setQuestionPoolingMode(questionPoolingMode: ORDERED) { assessmentId }

        # change the order
        switchQuestions(firstNumber: 1, secondNumber: 2) { assessmentId }

        # remove a question
        removeQuestion(number: 3) { assessmentId }

        addMultipleChoiceQuestion(input: {
            text: "What is the capital of France?"
            answers: [
                { text: "Paris", correct: true }
                { text: "London", correct: false }
                { text: "Rome", correct: false }
            ]
        }) { assessmentId }

        result: addMultipleChoiceQuestion(input: {
            text: "What is the capital of Italy?"
            answers: [
                { text: "Paris", correct: false }
                { text: "London", correct: false }
                { text: "Rome", correct: true }
            ]
        }) {
            assessmentId
            questionPoolingMode
            ...
        }
    }
}
```

### CODE REVIEW

File: src/main/resources/graphql/service/quiz.graphqls [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    """
+    text: String!
+    """
+    List of answers.
+    """
+    answers: [MultipleChoiceAnswer!]!
+    """
+    How many answers the user has to select. This is computed from the list of answers.
+    """
+    numberOfCorrectAnswers: Int!
```
Comment (SkyDiverStar) [inline src/main/resources/graphql/service/quiz.graphqls:97]: Is that automatically computed in the backend, when the frontend submits the answers? Or should the number of correct answers be first computed and than submitted in this field? 
Hard to grasp for me personally.

Comment (PaulBredl) [reply src/main/resources/graphql/service/quiz.graphqls:97]: With "This is computed from the list of answers" I meant in the backend.
This field is not present in the create or update inputs.

File: src/test/java/de/unistuttgart/iste/gits/quiz_service/matcher/MultipleChoiceAnswerDtoToEntityMatcher.java
Comment (PaulBredl) [inline src/test/java/de/unistuttgart/iste/gits/quiz_service/matcher/MultipleChoiceAnswerDtoToEntityMatcher.java:1]: Because comparing inputs to dtos, dtos to entities and entities to inputs was always ugly, I decided to do the proper solution and impleted matchers. They check for example if a QuestionEntity matches the corresponding CreateQuestionInput and so on

File: src/main/resources/graphql/service/query.graphqls [hunk 2]
```diff
@@ -0,0 +1,10 @@
+type Query {
+    """
+    Get quiz by assessment ID.
+    """
+    quizByAssessmentId(assessmentId: UUID!): Quiz!
```
Comment (myluki2000) [inline src/main/resources/graphql/service/query.graphqls]: Please provide a batching query for this

File: src/main/resources/graphql/service/mutation.graphqls [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+    """
+    Modify a quiz.
+    """
+    mutateQuiz(assessmentId: UUID!): QuizMutation!
+
+    """
+    Delete a quiz.
+    """
+    deleteQuiz(assessmentId: UUID!): UUID!
```
Comment (myluki2000) [inline src/main/resources/graphql/service/mutation.graphqls:15]: Can this be moved into the QuizMutation type? This'll make it easier later to implement authorization

**Review by myluki2000 — CHANGES_REQUESTED** [review]
Some schema changes pls

File: src/main/resources/graphql/service/mutation.graphqls (hunk 3 above)
Comment (PaulBredl) [reply src/main/resources/graphql/service/mutation.graphqls:15]: yes, but in the other service we don't even have this concept of the nested mutation. So how would we do it there?

Comment (PaulBredl) [reply src/main/resources/graphql/service/mutation.graphqls:15]: I think thematically it would make more sense if it would be outside of the QuizMutation

Comment (PaulBredl) [reply src/main/resources/graphql/service/mutation.graphqls:15]: This mutation should not be accessible from the outside at all I think. I can mark it as deprecated. Deletion should happen when the content is deleted

**Review by BlueWaves2 — APPROVED** [review]
Looks fine to me!
